### PR TITLE
ASTNode and basic types

### PR DESCRIPTION
**Pull request by pniedzwiedzinski**
In this PR I added `ASTNode`, `Runtime` and `Program` types. I also created temp type `PseudoValue` which will be a container for storing dynamic values but now it's just `int`.

I also added couple of sample resolvers to test it.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	PseudoValue value;
+	void (*resolver)(Runtime* r, ASTNode* self);
+	ASTNode* left;
+	ASTNode* right;
+	void resolve(Runtime* r) {
+		resolver(r, this);
+	}
+};
+
+void constIntResolver(Runtime* r, ASTNode* self) {
```
Comment (mhnowak) [inline src/main.cpp:9]: Why this needs to be a public method? Also it strictly uses ASTNode. Maybe just move it to this struct?

File: src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	r->acc = self->value;
+}
+
+ASTNode* createConstInt(PseudoValue value) {
+	return new ASTNode{value, &constIntResolver, nullptr, nullptr};
+}
+
+void sumIntResolver(Runtime* r, ASTNode* self) {
```
Comment (mhnowak) [inline src/main.cpp:17]: same with these. If we don't want these methods to be in struct think it'd be better to create a separate class for them could be even static methods if needed.

File: src/main.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	self->left->resolve(r);
+	PseudoValue leftValue = r->acc;
+	self->right->resolve(r);
+	PseudoValue rightValue = r->acc;
+	r->acc = leftValue + rightValue;
+}
+
+ASTNode* createSum(ASTNode* a, ASTNode* b) {
```
Comment (mhnowak) [inline src/main.cpp:25]: same as the above

File: src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		resolver(r, this);
+	}
+};
+
+void constIntResolver(Runtime* r, ASTNode* self) {
+	r->acc = self->value;
+}
+
+ASTNode* createConstInt(PseudoValue value) {
+	return new ASTNode{value, &constIntResolver, nullptr, nullptr};
```
Comment (mhnowak) [inline src/main.cpp:14]: Please add custom constructor for creating new object so you don't need to pass nullptr as arguments.

File: src/main.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+#define SUBSTRACT	'-'
+#define MULTIPLY	'*'
+#define DIVIDE		'/'
+
+//TODO: Dynamic typing, for now every value is int
+#define PseudoValue int // To enable dynamic typing we need to use one type for every value
+
+#define Program std::vector<ASTNode*>
+
+struct Runtime {
```
Comment (mhnowak) [inline src/main.cpp]: Also could be good to create separate files for each struct here. What do you think?

File: src/main.cpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
+#include <stdio.h>
+#include <vector>
+#include <iostream>
+
+#define ADD 		'+'
+#define SUBSTRACT	'-'
+#define MULTIPLY	'*'
+#define DIVIDE		'/'
```
Comment (mhnowak) [inline src/main.cpp]: why 2 tabs? use 1 please.

**Review by mhnowak — CHANGES_REQUESTED** [review]
Maybe it could be useful to create some abstract class like `resolver` To specify how resolvers should look like.

File: src/main.cpp (hunk 6 above)
Comment (pniedzwiedzinski) [inline src/main.cpp]:
```suggestion
#define ADD 	'+'
#define SUBSTRACT	'-'
#define MULTIPLY	'*'
#define DIVIDE	'/'
```

**Comment by pniedzwiedzinski** [comment]
> Maybe it could be useful to create some abstract class like `resolver` To specify how resolvers should look like.

This is just a temporary implementation to make a demo for ASTNode works. For sure we need a way to specify how resolvers should work, class sounds good. But I think it's outside the scope of this PR - so let's ignore it for now.

File: src/main.cpp (hunk 5 above)
Comment (pniedzwiedzinski) [reply src/main.cpp]: Sounds good :+1:

**Pull request merged**